_posts/2020-07-06-IP_address_manager.md

```diff
@@ -0,0 +1,179 @@
+---
+title: "Introducing the Metal3 IP Address Manager"
+date: 2020-07-06
+draft: false
+categories: ["metal3", "baremetal", "IPAM", "ip address manager"]
+author: Maël Kimmerlin
+---
+
+As a part of developing the Cluster API Provider Metal3 (CAPM3) v1alpha4
+release, the Metal3 crew introduced a new project : its own IP Address Manager.
+This blog post will go through the motivations behind such a project, the
+features that it brings, its use in Metal3 and the future work.
+
+## What is the IP Address Manager ?
+
+The IP Address Manager (IPAM) is a controller that provides IP addresses
+and manages the allocations of IP subnets. It is not a DHCP in that it only
+reconciles Kubernetes objects and does not answer to any DHCP queries. It
+allocates IP addresses on request, but does not handle any use of those
+addresses.
+
+This sounds like the description of any IPAM system, no ? Well the twist
+is that this manager is based on Kubernetes to specifically handle some
+constraints from Metal3. We will go through the different issues that this
+project tackles.
+
+When deploying nodes in bare metal environment, there are a lot of possible
+variations. This project specifically aims to solve the cases where static
+IP address configurations are needed. It is designed to specifically address
+this in the Cluster API (CAPI) context.
+
+CAPI addresses the deployment of Kubernetes clusters and nodes, using
+the Kubernetes API. As such, it uses objects such as Machine Deployments
+(similar to deployments for pods) that takes care of creating the requested
+number of machines, based on templates. The replicas can be increased by the
+user, triggering the creation of new machines based on the provided templates.
+This mechanism does not allow for flexibility to be able to provide static
+addresses for each machine. The manager adds this flexibility by providing
+the address right before provisioning the node.
+
+In addition, all the resources from the source cluster must support the CAPI
+pivoting, i.e. being copied and recreated in the target cluster. This means
+that all objects must contain all needed information in their spec field to
+recreate the status in the target cluster without losing information. All
+objects must, through a tree of owner references, be attached to the cluster
+object, for the pivoting to proceed properly.
+
+In a nutshell, the manager provides an IP Address allocation service, based
+on Kubernetes API and fulfilling the needs of Metal3, specifically the
+requirements of CAPI.
+
+## How does it work ?
+
+The manager follows the same logic as the volume allocation in Kubernetes,
+with a claim and an object created for that claim. There are three type of
+objects defined, the IPPool, the IPClaim and the IPAddress objects.
```

Comment:
I'm not sure if it gets rendered differently for the web site, but normally in markdown when referring to an exact type like this, I would do:

```
the `IPPool`, the `IPClaim`, and the `IPAddress` objects.
```

Reply:
fixed, thanks